### PR TITLE
Add Flex websockets sample

### PR DESCRIPTION
**Pull request by andrewsg**

### CODE REVIEW

File: appengine/flexible/websockets/app.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  python_version: 3
+
+# Use only a single instance, so that this local-memory-only chat app will work
+# consistently with multiple users. To work across multiple instances, an
+# extra-instance messaging system or data store would be needed.
+manual_scaling:
+  instances: 1
+
+
+# For applications which can take advantage of session affinity
```
Comment (ace-n) [inline appengine/flexible/websockets/app.yaml]: Is this necessary?

File: appengine/flexible/websockets/main.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2017 Google Inc.
```
Comment (ace-n) [inline appengine/flexible/websockets/main.py]: **2018** Google **LLC**?

Apply this comment throughout your PR, if you make this change.

File: appengine/flexible/websockets/main_test.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    check_server('http://{}/'.format(bind_to))
+
+    yield bind_to
+
+    server.kill()
+
+
+def test_http(server):
+    result = requests.get('http://{}/'.format(server))
```
Comment (ace-n) [inline appengine/flexible/websockets/main_test.py]: I take it we have to support Python 2.7?

If not, should we use f-strings here?

(Apply changes - if there are any - throughout your PR.)

File: appengine/flexible/websockets/templates/index.html [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+{# [START index] #}
+<!doctype html>
+<html>
+  <head>
+    <title>Google App Engine Flexible Environment - Python Websockets Chat</title>
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  </head>
+  <body>
+
+    <!-- [START form] -->
```
Comment (ace-n) [inline appengine/flexible/websockets/templates/index.html]: This is an awfully generic region tag, which makes it hard to track internally. :)

Is it possible to use something more self-explanatory? (e.g. `gae_websockets_form`?)

File: appengine/flexible/websockets/templates/index.html [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    <div>
+      <p>Status:</p>
+      <ul id="chat-status"></ul>
+    </div>
+    <!-- [END form] -->
+
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
+    <script>
+    // [START js]
```
Comment (ace-n) [inline appengine/flexible/websockets/templates/index.html]: Ditto - generic region tag name.

File: appengine/flexible/websockets/templates/index.html [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+         "wss" (WebSocket Secure) */
+      var scheme = window.location.protocol == "https:" ? 'wss://' : 'ws://';
+      var webSocketUri =  scheme
+                          + window.location.hostname
+                          + (location.port ? ':'+location.port: '')
+                          + '/chat';
+
+      /* Get elements from the page */
+      var form = $('#chat-form');
+      var textarea = $('#chat-text');
```
Comment (ace-n) [inline appengine/flexible/websockets/templates/index.html]: Nit: should this be moved into an included `.js` file, or is that too much work?

File: appengine/flexible/websockets/templates/index.html [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
+    <script>
+    // [START js]
+    $(function() {
+      /* If the main page is served via https, the WebSocket must be served via
+         "wss" (WebSocket Secure) */
+      var scheme = window.location.protocol == "https:" ? 'wss://' : 'ws://';
+      var webSocketUri =  scheme
+                          + window.location.hostname
+                          + (location.port ? ':'+location.port: '')
```
Comment (ace-n) [inline appengine/flexible/websockets/templates/index.html]: Unless we're worried about compatibility with older browsers - can we use JS template strings?

File: appengine/flexible/websockets/main.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+app = Flask(__name__)
+sockets = Sockets(app)
+
+
+@sockets.route('/chat')
+def chat_socket(ws):
+    while not ws.closed:
+        message = ws.receive()
+        if message is None:  # message is "None" if the client has closed.
```
Comment (ace-n) [inline appengine/flexible/websockets/main.py]: Nit: Even with line 28's check, I take it this is necessary? (i.e. I presume `ws.closed` won't be true until the last `ws.receive()` call is made.)

File: appengine/flexible/websockets/app.yaml (hunk 1 above)
Comment (andrewsg) [reply appengine/flexible/websockets/app.yaml]: The PM requested it be added, as a compromise after some debate about the feature.

File: appengine/flexible/websockets/main_test.py (hunk 3 above)
Comment (andrewsg) [reply appengine/flexible/websockets/main_test.py]: Flex supports both 3 and 2.7, so might as well leave it for compatibility reasons.

File: appengine/flexible/websockets/templates/index.html (hunk 6 above)
Comment (andrewsg) [reply appengine/flexible/websockets/templates/index.html]: For simplicity's sake (managing statics is another layer of complexity beyond templates) I'd say it's clearer as-is

File: appengine/flexible/websockets/templates/index.html (hunk 7 above)
Comment (andrewsg) [reply appengine/flexible/websockets/templates/index.html]: Sure, but I'm not familiar with them -- can you give an example?

File: appengine/flexible/websockets/main.py (hunk 8 above)
Comment (andrewsg) [reply appengine/flexible/websockets/main.py]: Yes, this is definitely necessary due to nuances on how receive() works in practice.

File: appengine/flexible/websockets/templates/index.html (hunk 4 above)
Comment (andrewsg) [reply appengine/flexible/websockets/templates/index.html]: Good call, fixed

File: appengine/flexible/websockets/templates/index.html (hunk 5 above)
Comment (andrewsg) [reply appengine/flexible/websockets/templates/index.html]: Fixed

File: appengine/flexible/websockets/main.py (hunk 2 above)
Comment (andrewsg) [reply appengine/flexible/websockets/main.py]: Fixed

**Comment by andrewsg** [comment]
PTAL @ace-n

File: appengine/flexible/websockets/templates/index.html (hunk 7 above)
Comment (ace-n) [reply appengine/flexible/websockets/templates/index.html]: Discussed offline - decided to keep it as-is.

**Comment by ace-n** [comment]
LGTM, assuming these region tags have been added to the sample tracker.

**Pull request merged**